File: backend/app/websocket/manager.py
from typing import Dict, Any, List
from datetime import datetime
from app.core.socket import sio


class SessionRoomManager:
    def __init__(self):
        # session_id -> room_state dict
        self.room_states: Dict[str, Dict[str, Any]] = {}
        # session_id -> { client_id -> client_info_dict }
        self.session_clients: Dict[str, Dict[str, Dict[str, Any]]] = {}
        # sid -> (session_id, client_id) mapping
        self.client_sessions: Dict[str, tuple[str, str]] = {}

    def get_room_state(self, session_id: str) -> Dict[str, Any]:
        if session_id not in self.room_states:
            self.room_states[session_id] = {
                "activePage": "inventory",
                "selectedTowerId": 1,
                "selectedUnitId": None,
                "galleryPreview": None,
                "videoPlayback": None,
                "bookingModal": None,
                "clientsCount": 0
            }
        return self.room_states[session_id]

    def _get_client_list_payload(self, session_id: str) -> List[Dict[str, Any]]:
        clients_dict = self.session_clients.get(session_id, {})
        client_list = []
        for cid, info in clients_dict.items():
            client_list.append({
                "clientId": cid,
                "name": info["name"],
                "browser": info.get("browser", "Browser"),
                "operatingSystem": info.get("operatingSystem", "OS"),
                "connectedAt": info.get("connectedAt", ""),
                "currentPage": info.get("currentPage", "inventory"),
                "status": info["status"],
                "isPaused": info["paused"]
            })
        return client_list

    async def broadcast_client_list(self, session_id: str):
        client_list = self._get_client_list_payload(session_id)
        # Connected count is count of Live or Mirroring Paused clients
        active_count = sum(1 for c in client_list if c["status"] != "Disconnected")
        if session_id in self.room_states:
            self.room_states[session_id]["clientsCount"] = active_count

        payload = {
            "clientsCount": active_count,
            "clients": client_list
        }
        await sio.emit("client:list_updated", payload, room=session_id)
        await sio.emit("room_clients_updated", {"clientsCount": active_count}, room=session_id)

    async def join_client(self, sid: str, session_id: str, client_id: str, browser: str = "Browser", os: str = "OS"):
        if not client_id:
            client_id = sid

        if session_id not in self.session_clients:
            self.session_clients[session_id] = {}

        # Remove old sid mapping if client_id reconnected with a new sid
        old_info = self.session_clients[session_id].get(client_id)
        if old_info and old_info.get("socket_id") and old_info["socket_id"] != sid:
            old_sid = old_info["socket_id"]
            if old_sid in self.client_sessions:
                del self.client_sessions[old_sid]

        self.client_sessions[sid] = (session_id, client_id)
        await sio.enter_room(sid, session_id)

        current_page = self.get_room_state(session_id).get("activePage", "inventory")

        if client_id in self.session_clients[session_id]:
            # Refreshed or reconnected client
            client_info = self.session_clients[session_id][client_id]
            client_info["socket_id"] = sid
            client_info["browser"] = browser or client_info.get("browser", "Browser")
            client_info["operatingSystem"] = os or client_info.get("operatingSystem", "OS")
            client_info["status"] = "Mirroring Paused" if client_info.get("paused") else "Live"
        else:
            # New client
            client_count = len(self.session_clients[session_id]) + 1
            client_info = {
                "clientId": client_id,
                "socket_id": sid,
                "name": f"Client {client_count}",
                "browser": browser or "Browser",
                "operatingSystem": os or "OS",
                "connectedAt": datetime.now().strftime("%I:%M %p"),
                "currentPage": current_page,
                "status": "Live",
                "paused": False
            }
            self.session_clients[session_id][client_id] = client_info

        state = self.get_room_state(session_id)
        # Emit initial room state to newly connected socket
        await sio.emit("room_state", state, to=sid)

        await self.broadcast_client_list(session_id)

    async def leave_client(self, sid: str):
        if sid in self.client_sessions:
            session_id, client_id = self.client_sessions.pop(sid)
            await sio.leave_room(sid, session_id)

            if session_id in self.session_clients and client_id in self.session_clients[session_id]:
                c_info = self.session_clients[session_id][client_id]
                # Only mark disconnected if this sid matches active socket for client_id
                if c_info.get("socket_id") == sid:
                    c_info["status"] = "Disconnected"
                    c_info["socket_id"] = None
                    await self.broadcast_client_list(session_id)

    async def update_and_broadcast(self, sid: str, event_name: str, payload: Dict[str, Any]):
        session_data = self.client_sessions.get(sid)
        if not session_data:
            return

        session_id, client_id = session_data
        state = self.get_room_state(session_id)

        # Update local room state cache
        if event_name == "sync_navigation":
            state["activePage"] = payload.get("activePage", "inventory")
            if session_id in self.session_clients and client_id in self.session_clients[session_id]:
                self.session_clients[session_id][client_id]["currentPage"] = state["activePage"]
                await self.broadcast_client_list(session_id)
        elif event_name == "sync_tower":
            state["selectedTowerId"] = payload.get("selectedTowerId")
        elif event_name == "sync_unit":
            state["selectedUnitId"] = payload.get("selectedUnitId")
        elif event_name == "sync_gallery":
            state["galleryPreview"] = payload.get("galleryPreview")
        elif event_name == "sync_video":
            state["videoPlayback"] = payload.get("videoPlayback")
        elif event_name == "sync_booking_modal":
            state["bookingModal"] = payload.get("bookingModal")

        # Selective broadcast: only emit to sockets of non-paused clients in this room (excluding sender)
        clients_in_room = self.session_clients.get(session_id, {})
        for cid, cinfo in clients_in_room.items():
            target_sid = cinfo.get("socket_id")
            if target_sid and target_sid != sid and not cinfo.get("paused") and cinfo.get("status") == "Live":
                await sio.emit(event_name, payload, to=target_sid)

    async def pause_client(self, session_id: str, client_id: str):
        if session_id in self.session_clients and client_id in self.session_clients[session_id]:
            cinfo = self.session_clients[session_id][client_id]
            cinfo["paused"] = True
            if cinfo["status"] != "Disconnected":
                cinfo["status"] = "Mirroring Paused"
            
            target_sid = cinfo.get("socket_id")
            if target_sid:
                await sio.emit("mirror_status_changed", {"isPaused": True}, to=target_sid)
                
            await self.broadcast_client_list(session_id)

    async def resume_client(self, session_id: str, client_id: str):
        if session_id in self.session_clients and client_id in self.session_clients[session_id]:
            cinfo = self.session_clients[session_id][client_id]
            cinfo["paused"] = False
            if cinfo.get("socket_id"):
                cinfo["status"] = "Live"

            target_sid = cinfo.get("socket_id")
            if target_sid:
                # Sync full current state to resumed client
                state = self.get_room_state(session_id)
                await sio.emit("room_state", state, to=target_sid)
                await sio.emit("mirror_status_changed", {"isPaused": False}, to=target_sid)

            await self.broadcast_client_list(session_id)

    async def disconnect_client(self, session_id: str, client_id: str):
        if session_id in self.session_clients and client_id in self.session_clients[session_id]:
            cinfo = self.session_clients[session_id][client_id]
            target_sid = cinfo.get("socket_id")
            
            cinfo["status"] = "Disconnected"
            cinfo["socket_id"] = None

            if target_sid:
                await sio.emit("client:disconnected_by_presenter", {
                    "message": "You have been disconnected from this presentation by the presenter."
                }, to=target_sid)
                await sio.disconnect(target_sid)

            await self.broadcast_client_list(session_id)


room_manager = SessionRoomManager()

File: kiosk-backend/src/modules/websocket/manager.ts
import { getIO } from '../../core/socket';

export interface RoomState {
  activePage: string;
  selectedTowerId: number;
  selectedUnitId: number | null;
  galleryPreview: any | null;
  videoPlayback: any | null;
  bookingModal: number | null;
  clientsCount: number;
}

export interface ClientInfo {
  clientId: string;
  socket_id: string | null;
  name: string;
  browser: string;
  operatingSystem: string;
  connectedAt: string;
  currentPage: string;
  status: 'Live' | 'Mirroring Paused';
  paused: boolean;
}

export class SessionRoomManager {
  private roomStates: Map<string, RoomState> = new Map();
  private sessionClients: Map<string, Map<string, ClientInfo>> = new Map();
  private clientSessions: Map<string, { sessionId: string; clientId: string }> = new Map();

  public getRoomState(sessionId: string): RoomState {
    if (!this.roomStates.has(sessionId)) {
      this.roomStates.set(sessionId, {
        activePage: 'inventory',
        selectedTowerId: 1,
        selectedUnitId: null,
        galleryPreview: null,
        videoPlayback: null,
        bookingModal: null,
        clientsCount: 0,
      });
    }
    return this.roomStates.get(sessionId)!;
  }

  public getClientSessions(): Map<string, { sessionId: string; clientId: string }> {
    return this.clientSessions;
  }

  private _getClientListPayload(sessionId: string) {
    const clientsMap = this.sessionClients.get(sessionId);
    if (!clientsMap) return [];
    
    const clientList: any[] = [];
    clientsMap.forEach((info) => {
      clientList.push({
        clientId: info.clientId,
        name: info.name,
        browser: info.browser || 'Browser',
        operatingSystem: info.operatingSystem || 'OS',
        connectedAt: info.connectedAt || '',
        currentPage: info.currentPage || 'inventory',
        status: info.status,
        isPaused: info.paused,
      });
    });
    return clientList;
  }

  public async broadcastClientList(sessionId: string): Promise<void> {
    const io = getIO();
    const clientList = this._getClientListPayload(sessionId);
    const activeCount = clientList.length;

    const roomState = this.getRoomState(sessionId);
    roomState.clientsCount = activeCount;

    const payload = {
      clientsCount: activeCount,
      clients: clientList,
    };

    io.to(sessionId).emit('client:list_updated', payload);
    io.to(sessionId).emit('room_clients_updated', { clientsCount: activeCount });
  }

  public async joinClient(
    sid: string,
    sessionId: string,
    clientId: string,
    browser: string = 'Browser',
    os: string = 'OS'
  ): Promise<void> {
    const io = getIO();
    const resolvedClientId = clientId || sid;

    if (!this.sessionClients.has(sessionId)) {
      this.sessionClients.set(sessionId, new Map());
    }

    const clientsMap = this.sessionClients.get(sessionId)!;

    // Remove old sid mapping if client reconnected with new sid
    const oldInfo = clientsMap.get(resolvedClientId);
    if (oldInfo && oldInfo.socket_id && oldInfo.socket_id !== sid) {
      this.clientSessions.delete(oldInfo.socket_id);
    }

    this.clientSessions.set(sid, { sessionId, clientId: resolvedClientId });
    
    // Join socket room
    const socket = io.sockets.sockets.get(sid);
    if (socket) {
      socket.join(sessionId);
    }

    const currentPage = this.getRoomState(sessionId).activePage || 'inventory';

    if (clientsMap.has(resolvedClientId)) {
      // Refreshed / reconnected client
      const clientInfo = clientsMap.get(resolvedClientId)!;
      clientInfo.socket_id = sid;
      clientInfo.browser = browser || clientInfo.browser || 'Browser';
      clientInfo.operatingSystem = os || clientInfo.operatingSystem || 'OS';
      clientInfo.status = clientInfo.paused ? 'Mirroring Paused' : 'Live';
    } else {
      // New client
      const clientCount = clientsMap.size + 1;
      const now = new Date();
      const timeStr = now.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
      
      clientsMap.set(resolvedClientId, {
        clientId: resolvedClientId,
        socket_id: sid,
        name: `Client ${clientCount}`,
        browser: browser || 'Browser',
        operatingSystem: os || 'OS',
        connectedAt: timeStr,
        currentPage: currentPage,
        status: 'Live',
        paused: false,
      });
    }

    const state = this.getRoomState(sessionId);
    if (socket) {
      socket.emit('room_state', state);
    }

    await this.broadcastClientList(sessionId);
  }

  public async leaveClient(sid: string): Promise<void> {
    const io = getIO();
    const sessionData = this.clientSessions.get(sid);
    if (sessionData) {
      const { sessionId, clientId } = sessionData;
      this.clientSessions.delete(sid);

      const socket = io.sockets.sockets.get(sid);
      if (socket) {
        socket.leave(sessionId);
      }

      const clientsMap = this.sessionClients.get(sessionId);
      if (clientsMap && clientsMap.has(clientId)) {
        const cInfo = clientsMap.get(clientId)!;
        if (cInfo.socket_id === sid) {
          clientsMap.delete(clientId);
          if (clientsMap.size === 0) {
            this.sessionClients.delete(sessionId);
          }
          await this.broadcastClientList(sessionId);
        }
      }
    }
  }

  public async updateAndBroadcast(sid: string, eventName: string, payload: any): Promise<void> {
    const io = getIO();
    const sessionData = this.clientSessions.get(sid);
    if (!sessionData) return;

    const { sessionId, clientId } = sessionData;
    const state = this.getRoomState(sessionId);

    if (eventName === 'sync_navigation') {
      state.activePage = payload.activePage || 'inventory';
      const clientsMap = this.sessionClients.get(sessionId);
      if (clientsMap && clientsMap.has(clientId)) {
        clientsMap.get(clientId)!.currentPage = state.activePage;
        await this.broadcastClientList(sessionId);
      }
    } else if (eventName === 'sync_tower') {
      state.selectedTowerId = payload.selectedTowerId;
    } else if (eventName === 'sync_unit') {
      state.selectedUnitId = payload.selectedUnitId;
    } else if (eventName === 'sync_gallery') {
      state.galleryPreview = payload.galleryPreview;
    } else if (eventName === 'sync_video') {
      state.videoPlayback = payload.videoPlayback;
    } else if (eventName === 'sync_booking_modal') {
      state.bookingModal = payload.bookingModal;
    }

    // Selective broadcast: emit to active non-paused sockets (excluding sender)
    const clientsMap = this.sessionClients.get(sessionId);
    if (clientsMap) {
      clientsMap.forEach((cinfo) => {
        if (cinfo.socket_id && cinfo.socket_id !== sid && !cinfo.paused && cinfo.status === 'Live') {
          io.to(cinfo.socket_id).emit(eventName, payload);
        }
      });
    }
  }

  public async pauseClient(sessionId: string, clientId: string): Promise<void> {
    const io = getIO();
    const clientsMap = this.sessionClients.get(sessionId);
    if (clientsMap && clientsMap.has(clientId)) {
      const cinfo = clientsMap.get(clientId)!;
      cinfo.paused = true;
      cinfo.status = 'Mirroring Paused';

      if (cinfo.socket_id) {
        io.to(cinfo.socket_id).emit('mirror_status_changed', { isPaused: true });
      }
      await this.broadcastClientList(sessionId);
    }
  }

  public async resumeClient(sessionId: string, clientId: string): Promise<void> {
    const io = getIO();
    const clientsMap = this.sessionClients.get(sessionId);
    if (clientsMap && clientsMap.has(clientId)) {
      const cinfo = clientsMap.get(clientId)!;
      cinfo.paused = false;
      if (cinfo.socket_id) {
        cinfo.status = 'Live';
        const state = this.getRoomState(sessionId);
        io.to(cinfo.socket_id).emit('room_state', state);
        io.to(cinfo.socket_id).emit('mirror_status_changed', { isPaused: false });
      }
      await this.broadcastClientList(sessionId);
    }
  }

  public async disconnectClient(sessionId: string, clientId: string): Promise<void> {
    const io = getIO();
    const clientsMap = this.sessionClients.get(sessionId);
    if (clientsMap && clientsMap.has(clientId)) {
      const cinfo = clientsMap.get(clientId)!;
      const targetSid = cinfo.socket_id;

      clientsMap.delete(clientId);
      if (clientsMap.size === 0) {
        this.sessionClients.delete(sessionId);
      }

      if (targetSid) {
        this.clientSessions.delete(targetSid);
        const socket = io.sockets.sockets.get(targetSid);
        if (socket) {
          socket.emit('client:disconnected_by_presenter', {
            message: 'You have been disconnected from this presentation by the presenter.',
          });
          socket.disconnect(true);
        }
      }

      await this.broadcastClientList(sessionId);
    }
  }

  public async broadcastUnitBooked(sessionId: string, unitId: number, customerName: string): Promise<void> {
    const io = getIO();
    const state = this.getRoomState(sessionId);
    state.selectedUnitId = null;
    state.bookingModal = null;

    const payload = {
      unitId,
      customerName,
      timestamp: String(unitId),
    };

    io.to(sessionId).emit('unit_booked', payload);
    io.emit('global_inventory_changed', payload);
  }
}

export const roomManager = new SessionRoomManager();
